File: leetcode/1_two_sum.go
//地址：https://leetcode.com/problems/two-sum/
//直观思路，两个for循环，O(n^2)解法，但略呆
//进阶思路，用map空间换时间即可
//难度: 一星
package main

func twoSum(nums []int, target int) []int {
	m, l := make(map[int]int), len(nums)
	for i:=0;i < l;i++ {
		m[nums[i]] = i
	}
	for i:=0;i < l;i++ {
		o := target - nums[i]
		if _, b := m[o];b && i!=m[o]{
			return []int{i, m[o]}
		}
	}
	return []int{-1, -1}
}
func main(){
	twoSum([]int, 1)
}

File: leetcode/1_two_sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> m;
        vector<int> ret;
        for(auto i=0;i<nums.size();i++) { m[nums[i]] = i;}
        for(auto i=0;i<nums.size();i++){
            int findNum = target - nums[i];
            if(m.find(findNum)!=m.end() && m[findNum] != i){
                ret.push_back(i);
                ret.push_back(m[findNum]);
                break;
            }
        }
        return ret;
    }
};